src/mono/mono/mini/interp/jiterpreter.c

```diff
@@ -0,0 +1,1082 @@
+	//  trace will have a branch block opened, because that determines whether calls and branches
+	//  will unconditionally abort the trace or not.
+	gboolean inside_branch_block = FALSE;
+
+	// We scan forward through the entire method body starting from the current block, not just
+	//  the current block (since the actual trace compiler doesn't know about block boundaries).
+	for (InterpInst *ins = bb->first_ins; (ins != NULL) && (ins != last_ins); ins = ins->next) {
+		int category = jiterp_should_abort_trace(ins, &inside_branch_block);
+		switch (category) {
+			case TRACE_ABORT: {
+				jiterpreter_abort_counts[ins->opcode]++;
+				return current_trace_length >= jiterpreter_minimum_trace_length;
+			}
+			case TRACE_IGNORE:
+				break;
+			default:
+				current_trace_length++;
+				break;
+		}
+
+		// Once we know the trace is long enough we can stop scanning.
+		if (current_trace_length >= jiterpreter_minimum_trace_length)
+			return TRUE;
+	}
+
+	return FALSE;
+}
+
+/*
+ * Insert jiterpreter entry points at the correct candidate locations:
+ * The first basic block of the function,
+ * Backward branch targets (if enabled),
+ * The next basic block after a call instruction (if enabled)
+ * To determine whether it is appropriate to insert an entry point at a given candidate location
+ *  we have to scan through all the instructions to estimate whether it is possible to generate
+ *  a suitably large trace. If it's not, we should avoid the overhead of the jiterpreter nop
+ *  instruction that would end up there instead and not waste any resources trying to compile it.
+ */
+void
+jiterp_insert_entry_points (void *_td)
+{
+	if (!jiterpreter_traces_enabled)
+		return;
+	TransformData *td = (TransformData *)_td;
+
+	// Insert an entry opcode for the next basic block (call resume and first bb)
+	// FIXME: Should we do this based on relationships between BBs instead of insn sequence?
+	gboolean enter_at_next = TRUE;
+
+	for (InterpBasicBlock *bb = td->entry_bb; bb != NULL; bb = bb->next_bb) {
+		// Enter trace at top of functions
+		gboolean is_backwards_branch = FALSE,
+			is_resume_or_first = enter_at_next;
+
+		// If backwards branches target a block, enter a trace there so that
+		//  after the backward branch we can re-enter jitted code
+		if (jiterpreter_backward_branch_entries_enabled && bb->backwards_branch_target)
+			is_backwards_branch = TRUE;
+
+		gboolean enabled = (is_backwards_branch || is_resume_or_first);
+		// FIXME: This scan will likely proceed forward all the way out of the current block,
+		//  which means that for large methods we will sometimes scan the same instruction
+		//  multiple times and waste some work. At present this is unavoidable because
+		//  control flow means we can end up with two traces covering different subsets
+		//  of the same method in order to handle loops and resuming
+		gboolean should_generate = enabled && should_generate_trace_here(bb, td->last_ins);
```

Comment:
The code around here operates on the wrong assumption that interp instructions are linked together from the first all the way to the last instruction which is wrong. Instructions are linked together only within the same basic block. `td->last_ins` is pretty much useless and I should probably get rid of it at some point since it is bad design (after an `interp_add_ins` we can access the added instruction from `td->last_ins` so we set data into it)